src/components/app/ServiceWorkerManager.js

```diff
@@ -66,75 +130,198 @@ class ServiceWorkerManager extends PureComponent<Props, State> {
     });
   }
 
-  componentDidMount() {
-    if (process.env.NODE_ENV === 'production') {
-      this._installServiceWorker();
+  _hasDatasourceProfile(): boolean {
+    const { dataSource } = this.props;
+
+    // We use a switch here, to make sure that when somebody adds a new
+    // dataSource, we'll get a flow error if they forget to update here. This is
+    // important because it's not obvious which case new datasources will fall
+    // into.
+    switch (dataSource) {
+      case 'none':
+        return false;
+      case 'from-file':
+      case 'from-addon':
+      case 'public':
+      case 'from-url':
+      case 'compare':
+      case 'local':
+        return true;
+      default:
+        throw assertExhaustiveCheck(dataSource);
     }
   }
 
-  componentDidUpdate() {
-    const { phase } = this.props;
-    const { installStatus } = this.state;
+  // This function checks whether the profile is fully loaded.
+  _isProfileLoadedAndReady(): boolean {
+    const { phase, symbolicationStatus } = this.props;
 
-    if (phase !== 'FATAL_ERROR') {
-      return;
+    if (phase !== 'DATA_LOADED') {
+      // Note we don't use a switch for the phase because it has a lot of
+      // different values and won't likely change often. Hopefully this comment
+      // won't age badly.
```

Comment:
Narrator voice: "But it did age badly"

Reply:
Actually I have no idea if it will age badly.

Reply:
You could coerce the check to provide a bit more safety `    if (phase !== ('DATA_LOADED': Phase)) {`